### PR TITLE
feat(rt-R5): streaming rewire — typed LlmStreamEvents + tool lifecycle over SSE

### PR DESCRIPTION
**Pull request by FisherXZ**
## Summary

R5 of the Runtime Solidification Sprint (plan: `docs/superpowers/plans/2026-07-14-r5-streaming-rewire-plan.md`). Replaces the single end-of-run answer flush in `/api/agent/stream` with typed re-emission of `LlmStreamEvent`s + tool lifecycle over the existing SSE channel:

- **True token streaming** when no `search_memory` call happens this run — the model's own `text_delta`s forward live to the chat bubble.
- **Citation safety preserved**: once any `search_memory` `tool_start` fires, subsequent text is held back and the authoritative, citation-checked `result.answer` is flushed once at the end. An invalid citation still never reaches the client.
- **Live tool indicator**: `data-tool-pending` part + `AssistantTurn.pendingTool` + "Running search_memory" pending row in the reasoning trace.

## Changes

- `src/lib/harness.ts` (flagged additive deviation into R2-owned file): optional `onAgentLoopEvent` on `RunAgentInput`; `onEvent` now built off `onStep || onAgentLoopEvent`, raw event forwarded first — no silent no-op when `onStep` is omitted.
- `src/lib/memory/answer.ts`: same additive passthrough on `AnswerWithMemoryInput`.
- `src/lib/ui-message-stream.ts`: one-shot `answer()` retired for `answerDelta` / `answerEnd` / `answerFlush` / `toolPending` (lazy text-start, append-or-fresh flush). AI SDK surface stays confined to this boundary file (model-boundary test green).
- `src/app/api/agent/stream/route.ts`: consumes both hooks; `searchCalledSoFar`/`streamedLive` gating per plan Judgment call #2. Gate couples to the literal tool name `search_memory` — safe while `memoryRegistry()` registers exactly one tool; commented in the route.
- `src/app/chat/stream.ts`, `ReasoningTrace.tsx`, `ChatClient.tsx`: `pendingTool` plumbing, cleared on step settle / meta.
- Accepted, documented edge case (plan Judgment call #3): pre-tool narration streams live and also reappears as the step's `thought` line — asserted by a dedicated route test, not suppressed.

## Tests

+23 tests (347 passed / 78 pre-existing legacy-sqlite failures / 1 todo; baseline 324/78/1, **zero new failures**), all TDD red→green:
- `tests/harness-agent-loop-event.test.ts` (new, 3) — includes the onAgentLoopEvent-without-onStep path from the eng review's must-fix #1
- `tests/memory-answer-agent-loop-event.test.ts` (new, 2)
- `tests/ui-message-stream.test.ts` (new, 6)
- `tests/agent-stream-route.test.ts` (+4) — live streaming, buffered flush, backward-compat, narration double-surfacing
- `tests/chat-stream.test.ts` (+3), `tests/components/chat-pieces.test.tsx` (+2)

## Gates

- `npx vitest run`: 347/78-pre-existing/1 todo, zero new failures
- `npx tsc --noEmit`: clean
- `npm run lint`: pre-existing `embed.ts` warning only
- `npm run build`: green, `/api/agent/stream` + `/chat` present
- `tests/model-boundary.test.ts`: pass
- Live probe (openai override): search path → 1 buffered text-delta + `data-tool-pending` + 0 invalidCitations, `/runs/2` renders; no-search path → **33 live text-deltas**, 0 tool calls

## Plan deviations (test mechanics only)

1. Plan's fake adapter `return {stopReason}` dropped — repo's adapter seam yields events only (matches `tests/harness.test.ts`).
2. Plan's `beforeEach(() => mock.mockReset())` one-liner replaced with a braced body: `mockReset()` returns the mock and vitest invokes a function returned from `beforeEach` as a cleanup hook — with no args — which crashed the mock at teardown.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added incremental assistant response streaming for faster visible output.
  * Added clear progress indicators showing which tool is currently running.
  * Improved response handling when memory search is used, with a final verified answer after the search completes.
  * Preserved streamed steps, metadata, and citations in the conversation view.

* **Bug Fixes**
  * Improved handling of pending tool states so progress indicators clear reliably when processing finishes.

* **Tests**
  * Added coverage for streaming, tool progress updates, memory search behavior, and backward compatibility.

<h3>Greptile Summary</h3>

This PR rewires the `/api/agent/stream` route from a single end-of-run flush to typed, incremental SSE events: live `text_delta` forwarding when no `search_memory` call has occurred, a `data-tool-pending` part for the live reasoning trace indicator, and a citation-checked `answerFlush` once any search tool fires.

- **`harness.ts` / `answer.ts`**: additive `onAgentLoopEvent` hook threaded alongside `onStep`; raw events forwarded before the existing `tool_end → onStep` collapse, with correct backward compat.
- **`ui-message-stream.ts`**: `answer()` replaced with `answerDelta` / `answerEnd` / `answerFlush` / `toolPending`; lazy text-part open via `ensureStarted`.
- **Route gating logic**: a real issue exists in the `streamedLive=true + searchCalledSoFar=true` branch — `answerFlush` appends the full authoritative answer to the still-open text part rather than replacing it, so the client's accumulated `turn.answer` becomes `[pre-tool narration][full answer]`. This contaminated string is rendered verbatim and is also threaded into multi-turn conversation history.

<h3>Confidence Score: 3/5</h3>

The no-search path and pure buffered path work correctly; the problematic path where the model narrates before calling search_memory produces a client answer that is the pre-tool thought prepended to the authoritative response, and that string is stored in conversation history for future turns.

The gating logic in the route has a real gap: once pre-tool text has streamed live and search_memory subsequently fires, answerFlush appends the citation-checked answer to the already-open text part instead of replacing it. Every multi-turn conversation where the model thinks aloud before searching will accumulate corrupted history. The rest of the PR — the harness hook plumbing, ui-message-stream writer, pendingTool lifecycle, and test suite — is clean and well-exercised.

src/app/api/agent/stream/route.ts — the branch covering streamedLive=true + searchCalledSoFar=true needs either a retract/reset mechanism or suppression of live streaming when a tool call is eventually made.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/app/api/agent/stream/route.ts | Introduces dual gating (searchCalledSoFar + streamedLive) for live vs. buffered text streaming; the branch covering pre-tool narration followed by a successful search concatenates narration text with the authoritative answer in the client's turn.answer, corrupting the displayed response and conversation history. |
| src/lib/harness.ts | Additive change: onAgentLoopEvent hook threaded alongside existing onStep with correct ordering (raw event fires first, then onStep collapse). onEvent handler is created if either hook is provided, preserving backward compat. |
| src/lib/ui-message-stream.ts | answer() replaced with answerDelta/answerEnd/answerFlush + toolPending; lazy text-start via ensureStarted is correct. answerFlush appends rather than resets when answerStarted=true, which is intentional for the hybrid streaming scenario. |
| src/lib/memory/answer.ts | Minimal passthrough change: onAgentLoopEvent forwarded verbatim to runAgent, no new logic introduced. |
| src/app/chat/stream.ts | pendingTool field added to AssistantTurn; applyChunk correctly sets it on data-tool-pending and clears it on data-step and data-meta. Logic is clean. |
| src/app/chat/ReasoningTrace.tsx | pendingTool prop wired correctly; pendingLabel derivation is straightforward and safe. |
| src/app/chat/ChatClient.tsx | pendingTool threaded from e.turn to ReasoningTrace; one-line change, correct. |
| tests/agent-stream-route.test.ts | Four new tests covering live streaming, buffered flush, backward compat, and the narration double-surfacing edge case. The narration test asserts the current (contaminating) behavior rather than guarding against it. |
| tests/ui-message-stream.test.ts | Six tests covering the new writer surface (answerDelta, answerEnd, answerFlush, toolPending); well-structured and complete. |
| tests/harness-agent-loop-event.test.ts | New integration tests with a real fake adapter verifying onAgentLoopEvent receives llm/tool_start/tool_end events and that onStep still fires correctly alongside it. |
| tests/memory-answer-agent-loop-event.test.ts | Two passthrough tests for answerWithMemory; correct use of braced beforeEach body to avoid the mockReset teardown crash. |
| tests/chat-stream.test.ts | Three new applyChunk tests for pendingTool lifecycle; all correct. |
| tests/components/chat-pieces.test.tsx | Two component tests verifying the pendingTool label rendering path; straightforward and accurate. |

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant C as ChatClient
    participant R as /api/agent/stream
    participant H as harness.ts (runAgent)
    participant L as agentLoop

    C->>R: "POST {question, history}"
    R->>H: "answerWithMemory({onStep, onAgentLoopEvent})"
    H->>L: "runAgentLoop({onEvent})"

    loop Turn 1 (no search yet)
        L-->>H: "AgentLoopEvent {type:llm, text_delta}"
        H-->>R: onAgentLoopEvent(llm.text_delta)
        Note over R: !searchCalledSoFar → answerDelta, streamedLive=true
        R-->>C: SSE text-delta (live token)
    end

    L-->>H: "AgentLoopEvent {type:tool_start, name:search_memory}"
    H-->>R: onAgentLoopEvent(tool_start)
    Note over R: searchCalledSoFar=true
    R-->>C: SSE data-tool-pending

    L-->>H: "AgentLoopEvent {type:tool_end}"
    H-->>R: "onStep({index, tool, thought, observation})"
    R-->>C: SSE data-step

    L-->>H: "AgentLoopEvent {type:llm, text_delta} post-search"
    H-->>R: onAgentLoopEvent(llm.text_delta)
    Note over R: searchCalledSoFar=true, gated out

    H->>R: "return result {answer, runId, status}"
    Note over R: streamedLive&&!searchCalledSoFar=false, result.answer=true
    Note over R: answerFlush — answerStarted already true, narration+answer concatenated
    R-->>C: SSE text-delta (full answer appended)
    R-->>C: SSE text-end
    R-->>C: SSE data-meta
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant C as ChatClient
    participant R as /api/agent/stream
    participant H as harness.ts (runAgent)
    participant L as agentLoop

    C->>R: "POST {question, history}"
    R->>H: "answerWithMemory({onStep, onAgentLoopEvent})"
    H->>L: "runAgentLoop({onEvent})"

    loop Turn 1 (no search yet)
        L-->>H: "AgentLoopEvent {type:llm, text_delta}"
        H-->>R: onAgentLoopEvent(llm.text_delta)
        Note over R: !searchCalledSoFar → answerDelta, streamedLive=true
        R-->>C: SSE text-delta (live token)
    end

    L-->>H: "AgentLoopEvent {type:tool_start, name:search_memory}"
    H-->>R: onAgentLoopEvent(tool_start)
    Note over R: searchCalledSoFar=true
    R-->>C: SSE data-tool-pending

    L-->>H: "AgentLoopEvent {type:tool_end}"
    H-->>R: "onStep({index, tool, thought, observation})"
    R-->>C: SSE data-step

    L-->>H: "AgentLoopEvent {type:llm, text_delta} post-search"
    H-->>R: onAgentLoopEvent(llm.text_delta)
    Note over R: searchCalledSoFar=true, gated out

    H->>R: "return result {answer, runId, status}"
    Note over R: streamedLive&&!searchCalledSoFar=false, result.answer=true
    Note over R: answerFlush — answerStarted already true, narration+answer concatenated
    R-->>C: SSE text-delta (full answer appended)
    R-->>C: SSE text-end
    R-->>C: SSE data-meta
```

</a>

<sub>Reviews (1): Last reviewed commit: ["feat(r5): live tool-name label in the re..."](https://github.com/fisherxz/sourcecado/commit/53caab18cb3b084a3af8b94e6250d348eda97d1e) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44381082)</sub>

> Greptile also left **2 inline comments** on this PR.

---
**Deferral note (added by supervisor):** `streamAgentTurn` capturePayloads/redactionMode parity remains deferred — now to **R6** (session persistence is the next payload consumer). Latent-only today: no live caller sets redactionMode. Tracked in progress.md carry-forwards.

### CODE REVIEW

**Comment by FisherXZ** [comment]
## Principal review — adversarial pre-merge

**Merge risk: low.** The slice is well-scoped, additive, TDD-first, and the citation-safety invariant holds in both design and a live probe. One true-positive gap (client-disconnect termination) is **fixed on this branch** with a regression test.

| # | Priority check | Verdict | Notes |
|---|---|---|---|
| 1 | Client disconnect seam | **TRUE POSITIVE — fixed on branch** | `request.signal` was never threaded into `answerWithMemory`, and AI SDK v6 `createUIMessageStream` wraps every write in `safeEnqueue` (swallows write-after-cancel), so `writer.write` never throws on disconnect → the R1/R2 abandoned-stream net (20c303f) was **unreachable** via the real SSE path. Loops ran to completion after the client left, burning credits, landing `succeeded`. Pre-existing (base had the same), but R5 is the first real streaming consumer, so closed here: thread `request.signal → answerWithMemory → runAgent` (loop/`streamAgentTurn` already fully support it). |
| 2 | Citation safety gating (mixed path) | **VERIFIED SAFE** | Gate holds: text before a `search_memory` streams live; once `searchCalledSoFar` closes (set on `tool_start` *before* `executeTool`), post-search text never streams — final answer flushed via the citation-checked `result.answer`. No-search path streams unchecked, but `verifyAnswerCitations` is a no-op without a search bundle in **both** streaming and JSON paths — no new exposure. |
| 3 | Carry-forward: `capturePayloads`/`redactionMode` parity | **TRUE — re-deferred to R6** | `streamAgentTurn` still always writes `request_json`/`response_json` (model-gateway.ts:518,546) with no capture gate, unlike `callModel`. Re-deferred in progress.md (R5 §, line 98) but **not mentioned in the PR body**. Latent only — no live caller sets `redactionMode`. Disposition: acceptable deferral; add a one-line note to the PR body. |
| 4 | Narration double-surfacing (plan must-fix #2) | **VERIFIED — accepted + tested** | Judgment call #3; asserted by `tests/agent-stream-route.test.ts` (narration streams live AND reappears as the step's `thought`). |
| 5 | R0 fixes intact | **VERIFIED — no regression** | `stream.ts:103` still throws on `!res.ok`; `ChatClient.tsx:61` still filters `!e.errored` out of history. |
| 6 | SSE wire discipline | **VERIFIED** | `drainSse` buffers the incomplete tail across reads and `try/catch`es `JSON.parse` (fragmentation-safe); 90s `RUN_TIMEOUT_MS` → `AbortController` → fetch signal intact. |

### Minor / follow-up (non-blocking)
- **Abort ledger label:** in the live disconnect probe, an aborted OpenAI-adapter fetch is bucketed as `error_type='model_error'` rather than `'aborted'` (provider error not recognized as an abort in `streamAgentTurn`'s catch). Cosmetic — the run still correctly terminates and fails; worth a follow-up to normalize the label.
- **`searchCalledSoFar` gates on the literal `"search_memory"` name** — safe today (single-tool registry) and already documented with a coupling comment in `route.ts`. TRUE BUT TRIVIAL until a second citation-bearing tool joins the registry.

### Verification
- **Live disconnect probe (openai / gpt-4o-mini):** client cut at 1.5s → run landed `failed` in **1503ms** vs **4600ms** for the full run → early termination confirmed (before the fix it would have run to completion and landed `succeeded`).
- **Live streaming probe:** no-search question → **5 separate `text-delta`** events (true token streaming); search question → `data-tool-pending` + exactly **1** buffered `text-delta` (citation-checked flush).
- `npx tsc --noEmit` clean · `npm test` **348 passed / 78 pre-existing sqlite bindings failures / 1 todo — zero new failures** · `npm run build` OK (both `/api/agent/stream` and `/chat` present) · `tests/model-boundary.test.ts` green · lint clean (only the pre-existing `embed.ts` known warning).

Fix pushed to `feat/rt-r5-streaming-rewire` as `d283f70`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: src/app/api/agent/stream/route.ts [hunk 1]
```diff
@@ -19,12 +21,37 @@ export async function POST(request: Request) {
...
+        }
+      },
     });
-    if (result.answer) writer.answer(result.answer);
+
+    if (streamedLive && !searchCalledSoFar) {
+      writer.answerEnd();
+    } else if (result.answer) {
+      writer.answerFlush(result.answer);
+    }
```
[the remark this answers is not in the document]
Comment (FisherXZ) [reply src/app/api/agent/stream/route.ts:64]: TRUE POSITIVE — fixed in `f035317`.

Verified against current code: in the mixed path where narration streams live first (`answerDelta` → `answerStarted=true` in `ui-message-stream.ts:42-45`) and `search_memory` then fires (`route.ts:46` sets `searchCalledSoFar=true`), a failed run with no `result.answer` matched neither finalization branch — `streamedLive && !searchCalledSoFar` is false and `result.answer` is falsy — so the `text-start` opened for the `answer` part was never closed with a `text-end`.

Root cause was a missing terminal case in the finalization, not the writer (`answerEnd()` is already idempotent, guarded by `answerStarted`). Fix adds an `else` branch that always closes any open answer part, so the SSE text part is well-formed on the failed-mid-run path too. All existing finalization cases are untouched. Verified with a new regression test asserting exactly one `text-end` on the (streamedLive, searchCalledSoFar, no-answer) path (`tests/agent-stream-route.test.ts`), plus `npx tsc --noEmit`, full suite (349 passed / 78 pre-existing sqlite-bindings failures / 1 todo — zero new), and `npm run build`.

[the remark this answers is not in the document]
Comment (FisherXZ) [reply src/app/api/agent/stream/route.ts:64]: TRUE POSITIVE — fixed in `f035317` (same root cause as the greptile P2 on this line). Added the fallback so `writer.answerEnd()` runs for every `streamedLive` path that ends without `result.answer`, including after `search_memory` has fired, while preserving `answerFlush(result.answer)` for completed answers. Verified with a new regression test (`tests/agent-stream-route.test.ts`) asserting exactly one `text-end` closes the narration part on the failed-after-search path; `tsc`, full test suite (zero new failures), and `build` all green.

[the remark this answers is not in the document]
Comment (FisherXZ) [reply src/app/api/agent/stream/route.ts:64]: Mechanism confirmed, but this is an explicitly documented and accepted design decision, not an overlooked bug — leaving as-is pending an owner call.

The trace is correct: pre-search narration streams live via `answerDelta` (`ui-message-stream.ts:42`), `answerFlush` appends `result.answer` as an additional `text-delta` without retracting the narration (`ui-message-stream.ts:52-56`), and the client concatenates every `text-delta` into `turn.answer` (`stream.ts:80-81`), which is both rendered (`ChatClient.tsx:136`) and threaded into multi-turn history (`ChatClient.tsx:65`).

This exact edge case is called out in the R5 plan, Judgment call #3 (`docs/superpowers/plans/2026-07-14-r5-streaming-rewire-plan.md:21`): narration emitted before a same-turn `search_memory` "will already have streamed live … and will visibly precede the final flushed answer in the same bubble — no retraction/clearing is implemented." It was deliberately accepted as rare (native tool-use models emit little-to-no preamble before a tool call — the live SSE probe in the principal review showed 0 pre-search deltas on a real search) and low-impact (worst case a short narration line ahead of the answer; never an unchecked citation, since citation-bearing text always goes through the buffered/flushed path). A dedicated test documents it (`tests/agent-stream-route.test.ts` — "accepts that pre-tool narration streams live AND reappears in the steps thought field").

One incremental point worth an owner call: the plan bounded the impact as a cosmetic prefix in the answer bubble, but the same concatenation also lands in `ConversationTurn` history for subsequent turns (`ChatClient.tsx:61-66`) — a second-order effect the rationale did not explicitly weigh. Still rare and bounded, but if retraction is ever wanted, the cheap clamp is client-side: reset `turn.answer` when a fresh authoritative flush begins, rather than changing the live-streaming contract. Surfacing to the branch owner rather than auto-fixing, since it would alter a contract that was deliberately shaped.

[the remark this answers is not in the document]
Comment (FisherXZ) [reply src/app/api/agent/stream/route.ts:64]: Owner decision (Fisher, relayed by supervisor): **accept as-is** per R5 plan Judgment call #3 — empirically rare (live probe: 0 pre-search deltas), cosmetic impact, and the history-threading angle is accepted at the same risk level. The client-side clamp is noted as an option for a future slice if real-world runs ever surface visible narration. Resolving.
